Bluetooth: Mesh: Add randomization to PB-ADV

Balaklaka opened the pull request:
Add randomization to PB-ADV to follow the shall statement regrading randomization and spacing Generic Provisioning PDUs and Acks (MshPRT v1.1, section 5.3.3).

alxelax commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -219,12 +226,15 @@ static void reset_adv_link(void)
 	if (atomic_test_bit(link.flags, ADV_PROVISIONER)) {
-		/* Clear everything except the retransmit and protocol timer
-		 * delayed work objects.
+		/* Clear everything except the transmit_delay, retransmit and
```
This is not clear to me. Why do we need to keep untouched these fields if link is reset?

Balaklaka replied:
We can not clear the delayable timers as they are only initialized in the init function. If they are cleared the work scheduler will crash. But the delayed timers is canceled in some lines above.

alxelax reviewed (commented):
LGTM just some questions.

PavelVPV reviewed (changes requested):
What about this ack:
https://github.com/zephyrproject-rtos/zephyr/blob/7951b506c5597ff3f8a61b193c472a13ce3e943c/subsys/bluetooth/mesh/pb_adv.c#L391-L398
?

Link Open, Ack and Close are also Generic Provisioning PDUs and the rule shall be applied to them as well.

Please also run PTS tests to verify that they still pass.

Balaklaka commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -86,6 +86,9 @@ struct pb_adv {
 		struct net_buf_simple *buf;
+
+		/* Random transmit delay timer */
+		struct k_work_delayable transmit_delay;
```
I rename and added the random word in the value, but I think it is important to keep the transmit part as that say something about what is delayed.

Commit:
Bluetooth: Mesh: Add randomization to PB-ADV

Add randomization to PB-ADV to follow the shall statement
regrading randomization and spacing Generic Provisioning
PDUs and Acks (MshPRT v1.1, section 5.3.3).

Signed-off-by: Ingar Kulbrandstad <ingar.kulbrandstad@nordicsemi.no>

Balaklaka commented:
I changed the solution for this after discussion with @PavelVPV about his comments.

PavelVPV commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -341,6 +341,16 @@ static bool schedule_send(struct bt_mesh_ext_adv *adv)
 	     adv->tags & BT_MESH_ADV_TAG_BIT_FRIEND) ||
 	    (CONFIG_BT_MESH_RELAY_ADV_SETS > 0 && adv->tags & BT_MESH_ADV_TAG_BIT_RELAY)) {
 		k_work_reschedule(&adv->work, K_NO_WAIT);
+	} else if ((IS_ENABLED(CONFIG_BT_MESH_PB_ADV)) && adv->tags & BT_MESH_ADV_TAG_BIT_PROV) {
```
This won't work. It should be this tag https://github.com/zephyrproject-rtos/zephyr/blob/a15ef7927b23088bf016966daa8879fa4a128a78/subsys/bluetooth/mesh/adv.h#L51

LingaoM commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -341,6 +341,16 @@ static bool schedule_send(struct bt_mesh_ext_adv *adv)
+		 */
+		uint8_t random_delay;
+
+		(void)bt_rand(&random_delay, sizeof(random_delay));
+		delta = k_uptime_delta(&timestamp);
+		k_work_reschedule(&adv->work, K_MSEC(20 + random_delay % 31 - delta));
```
can use timestamp as random value, since sending timestamps indeed random.

Just:
k_work_reschedule(work, (20 + (timestamp % 31)) - delta);

alxelax replied:
```
Note
While this function returns time in milliseconds, it does not mean it has millisecond resolution. The actual resolution depends on CONFIG_SYS_CLOCK_TICKS_PER_SEC config option.
```
Doesn't seem this is random.

LingaoM replied:
Can use Z_TIMEOUT_MS_TICKS?
This look like random?

alxelax replied:
I'm quite sure it is better to use suitable functionality to get random value. Mesh shouldn't depend on platform hidden implementations. I guess `Z_TIMEOUT_MS_TICKS` depends on platform. Also, if you have for some reason the application time synchronized behavior, it stops being random again. 
Talking about efficiency, it is not important in the place of the delay calculation. The behavior is going to be delayed in any case by design. Couple of tens instructions more or less will not impact at all on the final time (time is in tens of milliseconds).

mia-ko commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -186,7 +208,12 @@ static void free_segments(void)
 		}
 
-		bt_mesh_adv_unref(adv);
+		/* If buffer is waiting on being send, do not unref the buffer.
```
```suggestion
		/* If the buffer is waiting to be sent, do not unref the buffer.
```

Balaklaka commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -317,6 +351,180 @@ static void protocol_timeout(struct k_work *work)
+	uint16_t random_delay;
+	uint32_t curr_time;
+
+	if (k_mem_slab_alloc(&delayable_advs_buf_pool, (void **)&ctx, K_NO_WAIT)) {
```
If is not able to allocate it will return false and it will try to push out a message. And then try to allocate again. So it will not use ctx if it is not set.

Balaklaka commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -371,6 +578,10 @@ static void gen_prov_cont(struct prov_rx *rx, struct net_buf_simple *buf)
 		}
 
 		return;
+	} else if (rx->xact_id == next_transaction_id(link.rx.id) && link.tx.adv[0]) {
```
This can happend because the adv is delayed and the link can be closed down.
This was needed when having prov_clear_tx in gen_prov_start and gen_prov_cont which clear the tx buffer, and then the message waiting to go out was failing. This lead to that the callback in gen_prov_ack was not called and remote provisioning failed in sending provisioning start because of the prov_clear_tx prevented the callback in gen_prov_ack to be called. This was needed to make the tx_complete to be called in the rpr_srv.

alxelax replied:
Could you add this comment here? This is quite rare case that requires deep understanding of it.

PavelVPV commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -71,6 +71,20 @@ enum {
+	uint32_t fired_time;
+};
+
+K_MEM_SLAB_DEFINE(delayable_advs_buf_pool,
```
> Is any memory management required at all? There is only one sending packet at the same time

If I don't miss anything, acks can be send in parallel to regular PDUs in case an ack is lost by one role, but the other role started sending next PDU, e.g.:
https://github.com/zephyrproject-rtos/zephyr/blob/fe75bb950f45c1740225a21a2638a20080f4e967/subsys/bluetooth/mesh/pb_adv.c#L453-L461

I tend to think that every allocated adv buf requires only `sys_snode_t` entry which is consecutively attached to a single list. The `k_work_delayable` then pulls entries from this list one after another thus also adhering the order of the PDUs which you pointed out about in https://github.com/zephyrproject-rtos/zephyr/pull/65629/files#r1439302794.

Balaklaka replied:
I have tried looking if I could remove this, but I decided that it was best to keep it. I have removed the slab allocator.
The reason is this way everything goes through the same functions, and I do not have to do anything on tside for sending the ack and the unacked messages. For both of them I would have had to keep the message, callback and user_data in seperate data structures. I also would have had to some how kept a record of what messages for tx.adv we have sent and not sent. As they have to be kept for retransmissions. I figured out that this was the easyest way of keeping a record on what need to be sent.

PavelVPV commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -212,16 +239,23 @@ static void prov_clear_tx(void)
 	LOG_DBG("");
+
+	/* Flush all pending delayable advs. */
+	while (delayable_adv_ctx_push()) {
```
`delayable_adv_ctx_push` will send every buffer in the queue. But here we don't really want this.

Balaklaka commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -71,6 +71,20 @@ enum {
+
+K_MEM_SLAB_DEFINE(delayable_advs_buf_pool,
+		  sizeof(struct delayable_adv_ctx),
+		  CONFIG_BT_MESH_PB_ADV_DELAYABLE_ADV_COUNT,
```
I thought about this, but then I though that it need to be bigger than struct bt_mesh_adv * pointers as it can hold a ack at the same time.

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -231,6 +231,13 @@ config BT_MESH_PB_ADV_RETRANS_TIMEOUT
 	help
 	  Timeout value of retransmit provisioning PDUs.
 
+config BT_MESH_PB_ADV_DELAYABLE_ADV_COUNT
```
Do we need this configurable? How do people should figure out how many advs they need for the provisioning to work correctly? Will it make any difference if they use 5 instead of 4 buffers? Won't they brake the standard by decreasing this value (random delay is a shall statement in the spec)?

PavelVPV replied:
I don't. But I don't understand how users will figure out what value they need to set to be sure that the provisioning always works and doesn't violate the spec.

From what I see, the Transaction Start, Transaction Continuation and Link Open uses `link.tx.adv[]` to keep allocated adv buffers. All these messages don't use `bt_mesh_send_cb` callback as they use `send_reliable` function.

For the Link Ack, Link Close and Transaction Ack allocated adv buffer isn't stored anywhere. They also use `bt_mesh_send_cb` callback (Link Ack actually may not need the callback). All these 3 messages can't be sent at the same time: Link Ack is sent only when link is not opened yet (thus no Transaction PDUs can't be sent yet), Link Close is sent at the end of the provisioning (thus no Transaction PDUs can't be sent after that). This means we just need one extra pointer to adv buffer (+ `bt_mesh_send_cb` and `cb_data`). I see one corner case here: if we are sending Transaction Ack (or Link Ack) when triggering Link Close, but I don't see much problem with this.

If didn't miss anything, we just need extra adv buffer (+ `bt_mesh_send_cb` and `cb_data`) for Link Ack, Link Close and Transacation Ack messages and a `k_work_delayable` that checks if this extra adv buffer or `link.tx.adv[]` are set and sends them one by one with random delay.

One option to avoid storing `bt_mesh_send_cb` callback and `cb_data` is to change the advertiser api to allow to set the callback and the data when allocating an adv buffer.

Balaklaka replied:
I look at that I did not think it was a better idea. We need a way if keeping track of which link.tx.adv that has been sent so we need a seperate list than link.tx.adv, because we can not delete the link.tx.adv when it is sent because we need to keep it because of retransmission. So then we need a seperate mechanise of keeping track of that is going to be sent. For the ack and the sending the unack we need also to keep the pointers to the adv buffer and the callback and cb_data. So that has also to be made into seprate entries. 
I think it is best to keep this into one array/list for all of them. That makes it the easiest what to maintain and understand what it is use for. It can also be recognized from other parts of the code.

Balaklaka replied:
I do not see that what you suggest is a better solution.

PavelVPV commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -186,7 +212,12 @@ static void free_segments(void)
 		}
 
-		bt_mesh_adv_unref(adv);
+		/* If the buffer is waiting to be sent, do not unref the buffer.
```
Do we really need to send this segment if we aborting the transfer? Why we can't remove this buffer and reschedule the work?

Balaklaka replied:
This is not always called from reset. This can be called at point where it has messages waiting on going out. This way the tx buffer is cleared for the next transfer, but the messages waiting on going out will be sent. There is several places the prov_clear_tx/free_segments are called and in some way it thinks it is finished as it was before. But now it still has messages going out because of the delay.

PavelVPV replied:
But why the need to be sent? Looking at the logic above they are just dropped: https://github.com/zephyrproject-rtos/zephyr/blob/9074feabf6f827f6b37a69d77a5e13093799a905/subsys/bluetooth/mesh/pb_adv.c#L181-L189

PavelVPV commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -317,6 +359,142 @@ static void protocol_timeout(struct k_work *work)
+			return NULL;
+		}
+
+		ctx = delayable_adv_ctx_get_unused();
```
Previous implementation was always able to send provisioning PDUs, no we can't. Can we do something with this?

Balaklaka replied:
I do not understand what you mean by this?

alxelax replied:
If purge busy context, it should be able always to allocate new one. It should be able always to send provisioning PDU.

PavelVPV commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -317,6 +359,142 @@ static void protocol_timeout(struct k_work *work)
+	ctx = delayable_adv_ctx_get_unused();
+	if (!ctx) {
+		LOG_WRN("Purge pending delayed advertisement.");
+		if (!delayable_adv_ctx_push()) {
```
This brakes the shall statement of the spec regarding random delay. Can we allocate enough buffers to avoid this situation?

Balaklaka replied:
This was added as just in case we have 2 acks and a full tx buffer in the delay. Should we increase the size of buffer? This was added as a just in case mechanism, if we do not want it we can remove it, and then just return NULL. That will just make this resend the messages, that it could not send.

alxelax replied:
I do not think it breaks shall statement. It has already been delayed for some time. Unexpectedly completed random delay is still random. Unexpected completion is random act regarding protocol timings.

PavelVPV commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -317,6 +359,142 @@ static void protocol_timeout(struct k_work *work)
+	ctx = delayable_adv_ctx_assign(adv, cb, cb_data, retransmit);
+
+	if (!ctx) {
+		LOG_WRN("Unable to allocate delayable advertiser context, sending immediately.");
```
This brakes the shall statement of the spec regarding random delay. Can we allocate enough buffers to avoid this situation?

Balaklaka replied:
This was a solution to work around, this comment "https://github.com/zephyrproject-rtos/zephyr/pull/65629#discussion_r1439282764". What is best? Just drop the message and resend everything if retransmit is set, ack will not be sent if this happens.

Balaklaka commented on subsys/bluetooth/mesh/pb_adv.c:
```diff
@@ -317,6 +359,142 @@ static void protocol_timeout(struct k_work *work)
+
+	LOG_DBG("delay %u", random_delay);
+
+	k_work_reschedule(&link.random_delay, delay);
```
Should I then reintroduce the "uint32_t fired_time;" that was complained about was not needed. https://github.com/zephyrproject-rtos/zephyr/pull/65629#discussion_r1439302794 said that they should come in order, now they do come in order with a random delay between.

PavelVPV replied:
I think `k_work_schedule` would work here.
